File: creational/builder/builder.py

# Product class
class Product:
    def __init__(self):
        self._part_a = None
        self._part_b = None

    def set_part_a(self, part_a):
        self._part_a = part_a

    def set_part_b(self, part_b):
        self._part_b = part_b

    def show(self):
        print(f"Product built with PartA: {self._part_a}, PartB: {self._part_b}")

# Builder interface
class Builder:
    def build_part_a(self):
        pass

    def build_part_b(self):
        pass

    def get_result(self):
        pass

# Concrete Builder
class ConcreteBuilder(Builder):
    def __init__(self):
        self._product = Product()

    def build_part_a(self):
        self._product.set_part_a("PartA1")

    def build_part_b(self):
        self._product.set_part_b("PartB1")

    def get_result(self):
        return self._product

# Director class
class Director:
    def construct(self, builder):
        builder.build_part_a()
        builder.build_part_b()

if __name__ == "__main__":
    director = Director()
    builder = ConcreteBuilder()

    director.construct(builder)
    product = builder.get_result()
    product.show()

File: creational/builder/Builder.cs

using System;

// Product class
public class Product
{
    private string _partA;
    private string _partB;

    public void SetPartA(string partA)
    {
        _partA = partA;
    }

    public void SetPartB(string partB)
    {
        _partB = partB;
    }

    public void Show()
    {
        Console.WriteLine($"Product built with PartA: {_partA}, PartB: {_partB}");
    }
}

// Builder interface
public interface IBuilder
{
    void BuildPartA();
    void BuildPartB();
    Product GetResult();
}

// Concrete Builder
public class ConcreteBuilder : IBuilder
{
    private Product _product = new Product();

    public void BuildPartA()
    {
        _product.SetPartA("PartA1");
    }

    public void BuildPartB()
    {
        _product.SetPartB("PartB1");
    }

    public Product GetResult()
    {
        return _product;
    }
}

// Director class
public class Director
{
    public void Construct(IBuilder builder)
    {
        builder.BuildPartA();
        builder.BuildPartB();
    }
}

// Client code
class Program
{
    static void Main()
    {
        Director director = new Director();
        IBuilder builder = new ConcreteBuilder();

        director.Construct(builder);
        Product product = builder.GetResult();
        product.Show();
    }
}
